[dagster-airlift] Migration (#25822)
dagster-io/dagster: docs/content/integrations/airlift/tutorial/migrate.mdx

# Airflow Migration Tutorial: Migrating Assets

Previously, we completed the ["Observe" stage](/integrations/airlift/tutorial/observe) of the Airflow migration process by encoding the assets that are produced by each task. We also introduced partitioning to those assets.

Up until now, we have not touched the Airflow code base at all. Now it's time to begin the actual migration process.

> Once you have created corresponding definitions in Dagster to your Airflow tasks, you can proxy execution to Dagster on a per-task basis while Airflow is still controlling scheduling and orchestration. Once a task has been proxied, Airflow will kick off materializations of corresponding Dagster assets in place of executing the business logic of that task.

Reviewer: ```suggestion
Once you have created the corresponding definitions in Dagster for your Airflow tasks, you can proxy execution to Dagster on a per-task basis while Airflow still controls scheduling and orchestration. Once a task has been proxied, Airflow will kick off materializations of corresponding Dagster assets instead of executing the business logic of that task.
```

```yaml file=../../experimental/dagster-airlift/examples/tutorial-example/tutorial_example/airflow_dags/proxied_state/rebuild_customers_list.yaml
tasks:
  - id: load_raw_customers
    proxied: False
  - id: build_dbt_models
    proxied: False
  - id: export_customers
    proxied: False
```

Next, you will need to modify your Airflow DAG to make it aware of the proxied state. This is already done in the example DAG:

```python file=../../experimental/dagster-airlift/examples/tutorial-example/tutorial_example/snippets/dags_truncated.py
# Dags file can be found at tutorial_example/airflow_dags/dags.py
from pathlib import Path

from airflow import DAG
from dagster_airlift.in_airflow import proxying_to_dagster
from dagster_airlift.in_airflow.proxied_state import load_proxied_state_from_yaml

dag = DAG("rebuild_customers_list", ...)

...

# Set this to True to begin the proxying process
PROXYING = False

if PROXYING:
    proxying_to_dagster(
        global_vars=globals(),
        proxied_state=load_proxied_state_from_yaml(Path(__file__).parent / "proxied_state"),
    )
```

Set `PROXYING` to `True` or eliminate the `if` statement.

> The DAG will now display its proxied state in the Airflow UI. (There is some latency as Airflow evaluates the Python file periodically.)

Reviewer: ```suggestion
The DAG will now display its proxied state in the Airflow UI.
```

I think it's ok to forgo the latency warning here

<p align="center">

<Image
alt="Migration state rendering in Airflow UI"
src="/images/integrations/airlift/state_in_airflow.png"
width={528}
height={102}
/>

</p>

### Migrating individual tasks

In order to proxy a task, you must do two things:

> 1. First, ensure all associated assets are executable in Dagster by providing asset definitions in place of bare asset specs.

Reviewer: This may confused some people.

If by default we create AssetSpecs representing tasks through Airlift, maybe this can be rephrased to indicate that an asset must be defined for the task implementing the logic of that Airflow task.

I think we still need to do a better job of teaching people asset specs vs assets in our docs.
